### PR TITLE
perf: add annotations for non-concurrent cross-scoped duration measurements

### LINKED ISSUE #13860: Extend PerfAnnotationContext with begin and end events.
*Title*: *Extend `PerfAnnotationContext` with begin and end events.*

*Description*:
The current interface of PerfAnnotationContext allows easy recording of durations for a region of code inside a single lexical scope with `PERF_OPERATION` and `PERF_RECORD` annotations. But sometimes it might be useful to record durations between events happening in different scopes, e.g. how long it takes to destroy an `ActiveTcpConnection` after its corresponding `ActiveStream` has been destroyed (given they are 1-to-1).

I'd like to propose two additional annotations `PERF_BEGIN(category, description)` and `PERF_END(category, description)` asserting that a closing `PERF_END` always happens after `PERF_BEGIN`.

### PR DESCRIPTION
**Pull request by rojkov**
Commit Message: perf: add annotations for non-concurrent cross-scoped duration measurements
Additional Description:
Risk Level: Low
Testing: updated unit tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Fixes #13860

### CODE REVIEW

File: source/common/common/perf_annotation.cc
Comment (antoniovicente) [inline source/common/common/perf_annotation.cc:37]: RELEASE_ASSERT use seems risky.  Are these begin/end annotations meant to be used across event loop events or only within a single call stack?

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: Within a single call stack only unfortunately. Would be nice to have uniquely identifiable events in the event loop though. Is it possible?

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: I think it's OK to use RELEASE_ASSERT as this code will only be activated when the binary is compiled with perf annotation enabled.

And it wouldn't do any good to use ASSERT because you generally would want opt-builds for measuring performance.

File: source/common/common/perf_annotation.cc
Comment (jmarantz) [inline source/common/common/perf_annotation.cc:46]: might as well include the category & description in this message and the RELEASE_ASSERT above.

File: source/common/common/perf_annotation.h
Comment (jmarantz) [inline source/common/common/perf_annotation.h:47]: s/this/these/
s/makes/make/

File: source/common/common/perf_annotation.h
Comment (jmarantz) [inline source/common/common/perf_annotation.h:49]: This in reference only to a particular combination of category and description, right?

File: test/common/common/perf_annotation_test.cc
Comment (jmarantz) [inline test/common/common/perf_annotation_test.cc:32]: Can you also add begin/end checking with same-category different-description and vice versa?

File: source/common/common/perf_annotation.h
Comment (jmarantz) [inline source/common/common/perf_annotation.h:131]: seems like you don't need to repeat all of this paragraph from above, but could just reference the doc in begin().

**Review by jmarantz — COMMENTED** [review]
nice. Just some trivial comments.

File: source/common/common/perf_annotation.cc
Comment (antoniovicente) [reply source/common/common/perf_annotation.cc:37]: There's a better way to do all of this: have an RAII object that is kept around for the duration of the perf operation and keeps the necessary state instead of keeping the state on a global as a string in a map.

It nicely solves the issue of concurrent performance measurements and works across call stacks.

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: Yeah for cases where RAII is OK, that works today with `PERF_OPERATION` which is just a wrapper that has all the code compile to nothing when perf annotations are not compiled in. I think that should be preferred when feasible to use.

I assume that PERF_BEGIN and PERF_END are where the plumbing needed to hold an object seems not worth it, and correlating by string seems worth it. It doesn't look to me like this adds any overhead when not used, so that's why I'm OK with this if it makes annotations easier to implement.

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: My bad, I should have given more thought to the question: it may well happen that a "begin" and its corresponding "end" are called for different event loop events e.g. if you put them in ActiveTcpConnection's constructor and destructor respectively.

Comment (antoniovicente) [reply source/common/common/perf_annotation.cc:37]: I think that hitting this RELEASE_ASSERT is very likely if the performance measurement ever remains active at the time you unroll the stack to the event loop.  It just takes a second object of that type being benchmarked at the same time.

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: That's true. Applicability of these annotations is quite limited. I mentioned it the doc strings.

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: Sorry I re-read the class and realize that `PerfOperation` does not quite meet the RAII goals @antoniovicente had in mind. That's because in the use-case I built it for, I wanted to start an operation timer but record different categories depending on the outcome of an operation.

A subclass of `PerfOperation` and associated macros could be added where we commit to the category and description at construction time, and the record() happens automatically upon destruction. That would be safer to use than PERF_BEGIN and PERF_END.

@rojkov would that meet your needs? Or do you need the singleton to hold the string-pair map.

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: A RAII based timer is good to measure time spent for a function or a code block. In my case I wanted to measure how exactly wall time is spent during an HTTP request (actually a sequence of many HTTP requests): from establishing a connection to stream creation, from stream creation to headers decoding in a filter, etc... For this I had to maintain a map in the singleton.

Basically I wanted to have a very simple tool for basic [Off-CPU Analysis](http://www.brendangregg.com/offcpuanalysis.html) without resorting to Off-CPU sampling.

**Review by jmarantz — DISMISSED** [review]
@envoyproxy/senior-maintainers

File: source/common/common/perf_annotation.cc
Comment (mattklein123) [reply source/common/common/perf_annotation.cc:37]: I haven't fully paged in this conversation, but at minimum doesn't this have to be locked? I'm confused how there wouldn't be race conditions here. Should't we potentially use locking and also allow multiple concurrent measurements that then get added/closed at the end of the sequence? This could be handled by an RAII like handle like @antoniovicente is suggesting?

/wait-any

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: RE locking: yeah that is needed in general; good catch. It might not have been needed by @rojkov 's first usage. Note the ifdef PERF_THREAD_SAFE dictating whether we instantiate and hold the mutexes during map operations. The annotations I had added were (a) on operations that strictly occurred on the main thread and (b) were quick enough that taking the mutex distorted the results a little, hence the ifdef. We should make a PERF_THREAD_SAFE-version of the new map too.

RE RAII, I will let @rojkov describe the use-case in more detail but I thought in some case if you want to start an operation in one scope and end it in another you'd have to manage transfer for the RAII object, and that might be more intrusive to the code you're trying to measure. Using maps allows you to surgically inject some perf visibility without having to have a path to make the object visible in both the place where you are starting and ending the measurement. Having said that, it is not immediately obvious to me why you'd want to measure performance in that scenario.

Comment (mattklein123) [reply source/common/common/perf_annotation.cc:37]: The RAII part doesn't matter, but for this to be useful it seems like we need to support concurrent measurements that might run on multiple workers at the same time?

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: That's a good point; to make this work maybe there could be a third element to the hash-key, which would be a unique ID used for matching in `timestamps_map_`, but which is not used in `duration_stats_map_`, which aggregates across duplicate entries.

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: This PR is a third version of what was done initially. In the first version I mindlessly protected `timestamps_map_` updates with a mutex. [Then](https://github.com/rojkov/envoy/tree/perf-annotations-thread-aware) I realized that the hash-key should better be extended with `std::thread::id` of the current thread.

Finally I stumbled upon what @antoniovicente mentioned earlier: even if there's only one thread it's still possible that a sequence started in e.g. `decodeHeaders()` and stopped in `encodeTrailers()` may stay open waiting for a cluster response when a new connection to the listener triggers an event loop event leading to `PERF_BEGIN()` inside `decodeHeaders()` trying to update the already existing key in the map. That's why I decided to give up thread safety as not really needed for my use case and added some words of warning to the doc strings.

But it would great to extend the hash-key with a third element which could be used universally for matching. I don't know what that could be. Not thread ID.

Comment (mattklein123) [reply source/common/common/perf_annotation.cc:37]: I think the only way it can sensibly work is for the begin API to return a handle (with an internal incrementing ID or whatever) that is used to track the time. In practice, nothing actually has to be added to the perf system until the begin is closed with the end (kind of like the Timespan API we already have). This obviously brings us back to an RAII object that would have to be stored, but IMO this is probably fine as I think with appropriate ifdefs everything can probably be compiled out as a member of an owning object?

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: Yes, the `begin()` could store a `Timespan` object in a global queue then the corresponding `end()` would pop and `complete()` it. Theoretically `end()` might happen to pop a wrong object and that would distort StdDev, but Avg would be correct still.

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: Yeah that imprecision might be OK. But I still think including an extra ID timestamps_map_ may be practical. For example, if you are measuring performance spanning filter callbacks, you could use `this` as your extra ID. The cleanest way though is just for begin() to return an object that you store in your filter undef ifdef; then you don't need a map at all, or to worry about thread safety.

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: > The cleanest way though is just for begin() to return an object that you store in your filter undef ifdef

A-ha, now I see. In this case the owning object (e.g. filter) needs to be annotated too. And it has to be the same for a pair of begin() and end(). Right?

Comment (jmarantz) [reply source/common/common/perf_annotation.cc:37]: Yeah though I think the logical next step of what I suggested is that rather than calling begin() you just instantiate an RAII object yourself, throwing it in an ifdef'd member variable in (say) the filter; maybe via unique_ptr, and destructing it works like end().

Comment (rojkov) [reply source/common/common/perf_annotation.cc:37]: Understood, thanks!

Comment (mattklein123) [reply source/common/common/perf_annotation.cc:37]: Yes, exactly, ^ is what I'm proposing. I think we can make it so it compiles to nothing if annotations are not enabled.

Comment (rojkov) [reply source/common/common/perf_annotation.cc]: Updated. Turned out it's enough to add a few new macros to achieve the same result.

**Review by jmarantz — APPROVED** [review]
nice!

**Comment by mattklein123** [comment]
Can you check coverage?

/wait

**Comment by rojkov** [comment]
> Can you check coverage?

@mattklein123 Looks it was a CI glitch. Everything is Ok now.

**Review by mattklein123 — APPROVED** [review]
Awesome!

**Pull request merged**
